cs3342_09-master/README.md

```diff
@@ -0,0 +1,191 @@
+So is this: rrr
+
+This is not: rgb
+This is not: rggr    (it must be an odd length)
+
+## Q1.1  (5 points)
+
+Write a Chomsky type 2 grammar that describes any valid path containing at
+least one paver. Use `S` as the start state, and `r`, `g`, `b` as terminals that
+represent the tiles.
+
+## A1.1
+
+
+S -> color1 | color2 S color2
+
+color -> r|g|b
+
+## Q1.2  (1 point for the O() answer, 2 for the sentence)
+
+Using big-O notation, what is the likely memory requirement for a parser that
+can validate a particular path configuration, where `n` is the number of tiles?
+In one sentence, explain why.
+
+## A1.2
+
+O(1) becuase the parser would only need to iterate over the path one time 
+
+
+# Q2
+
+A simple sentence structure might be
+
+    "the" (zero or more adjectives) noun verb (optional adverb)
+
+Adjectives are "lazy" or "smelly"
+
+Nouns are "dog" or "cat"
+
+Verbs are "ate" and "ran"
+
+Adverbs are "slowly" and "noisily"
+
+The following are examples of valid sentences:
+
+* The smelly dog ran.
+* The smelly dog ran slowly.
+* The cat ate noisily.
+
+## Q2.1 (7 points)
+
+Write the BNF (not EBNF) description for this language.
+
+## A2.1
+
+\<sentence> ::= "the" + \<adj> + \<noun> + \<verb> + \<adverb>
+\<adj> ::= \<<empty>> | "lazy" | "smelly"
+\<adverb> ::=  \<<empty>> | "slowly" | "noisily"
+\<noun> ::=   "dog" | "cat"
+\<verb> ::=   "ate" | "ran"
+
+
+## Q2.2 (5 points)
+
+Write this grammar using EBNF with common extensions
+
+## A2.2
+
+\<sentence> ::= "the" + {\<adj>} + \<noun> + \<verb> + [\<adverb>]
+\<adj> ::= "lazy" | "smelly
+\<adverb> ::= "slowly" | "noisily"
+\<noun> ::=   "dog" | "cat"
+\<verb> ::=   "ate" | "ran"
+
+
+## Q2.3 (6 points)
+
+  Draw a diagram for an FSM which recognizes these sentences. Use EOI as the
+  event that occurs at the end-of-input. Start with a state named S0. Receiving
+  the word "the" in that state will transition to state S1. Name the final state
+  END.
+
+  You can hand draw it and snap a picture, or use a drawing tool. Either way,
+  upload the image, and put the file name in the answer below.
+
+  (Hint: my answer has seven states including the start and end states)
+
+
+## A2.3
+
+Name = A2.3
+
+
+
+## Q2.4 (6 points)
+
+Convert this diagram into a table of the form:
+
+Current state | Next word | Next state
+--------------|-----------|-----------
+    S0        |    the    |     S1
+    S1        |   . . .   |   . . .
+
+(hint: my version has 13 entries. Yours _might_ be different)
+
+## A2.4
+
+Current state | Next word | Next state
+--------------|-----------|-----------
+    S0        |    the    |     S1
+    S1        |   lazy    |     S2
+    S1        |   smelly  |     S2
+    S1        |   "empty" |     S2
+    S2        |   dog     |     S3
+    S2        |   cat     |     S3
+    S3        |   ran     |     S4
+    S3        |   ate     |     S4
+    S4        | slowly    |     S5
+    S4        | noisily    |     S5
+    S5        | EOI       |     END
```

Comment:
The diagram is fine past S2 (but you're missing the transition `S4 -> EOI -> END`.

In a state machine you have states (the circles) and directed transitions between the circles. The transitions are each labelled with an event. When you are in a particular state and you receive an event, you look for the transition with that even as a label and follow it to the next state.So every transition must be caused by an event, and «empty» is not an event. If it were, how could the FSM choose what to do in S1.

I was looking for something like

![image](https://user-images.githubusercontent.com/10648/78714446-87d3c980-78e1-11ea-91d2-d99055d1e86b.png)

# -3